internal/codecutil/encoding.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (C) MongoDB, Inc. 2023-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package codecutil
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsoncodec"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+)
+
+var ErrNilValue = errors.New("value is nil")
+
+// MarshalError is returned when attempting to transform a value into a document
+// results in an error.
+type MarshalError struct {
+	Value interface{}
+	Err   error
+}
+
+// Error implements the error interface.
+func (e MarshalError) Error() string {
+	return fmt.Sprintf("cannot transform type %s to a BSON Document: %v",
+		reflect.TypeOf(e.Value), e.Err)
+}
+
+// EncoderFn is used to functionally construct an encoder for marshaling values.
+type EncoderFn func(io.Writer) (*bson.Encoder, error)
+
+// MarshalValue will attempt to encode the provided value with the registry and
+// encoder function. If the encoder function does not exist, then this
+func MarshalValue(val interface{}, registry *bsoncodec.Registry, encFn EncoderFn) (bsoncore.Value, error) {
```

Comment:
Optional: You can further simplify this function by expecting the `Registry` to already be incorporated into the `Encoder` returned by the `EncoderFn`. Consider removing the `registry` param and embedding it into the `EncoderFn` before calling `MarshalValue`.